EtherKit/Models/TypedData.swift

```diff
@@ -0,0 +1,394 @@
+      }
+    case let .array(_, value):
+      var data = Data()
+      for subVal in value {
+        let subData = try subVal.encodeData(types: types)
+        data.append(subData)
+      }
+      return data
+    case let .object(type, value):
+      let typeHash = try types.getTypeHash(for: type)
+      let encodedData = try value.encodeData(typeName: type, types: types)
+      return (typeHash + encodedData).sha3(.keccak256)
+    }
+  }
+}
+
+public struct MessageObject {
+  public var values: [String: MessageValue] = [:]
+
+  public init(from object: [String: Any], typeName: String, types: Types) throws {
+    guard let type = types[typeName] else {
+      throw EtherKitError.web3Failure(reason: .parsingFailure)
+    }
+    for prop in type.properties {
+      guard let valueObject: Any = object[prop.name] else {
+        throw EtherKitError.web3Failure(reason: .parsingFailure)
+      }
+      values[prop.name] = try MessageValue.getMessageValue(from: valueObject, type: prop.type, types: types)
+    }
+  }
+
+  public func encodeData(typeName: String, types: Types) throws -> Data {
+    guard let mainType = types[typeName] else {
+      fatalError()
+    }
+
+    var data = Data()
+    for prop in mainType.properties {
+      guard let subVal = values[prop.name] else {
+        fatalError()
+      }
+      guard let subData = try? subVal.encodeData(types: types) else {
+        fatalError()
+      }
+      data.append(subData)
+    }
+    return data
+  }
+}
+
+public struct TypedData: ValueType {
+  public var types: Types
+  public var domain: Domain
+  public var primaryType: String
+  public var message: MessageObject
+
+  public static func value(from object: Any) throws -> TypedData {
+    guard let valueMaps = object as? [String: Any],
+      let typeDict: [String: Type] = try? valueMaps.value(for: "types") else {
+      throw EtherKitError.web3Failure(reason: .parsingFailure)
+    }
+    let types: Types = Types(types: typeDict)
+
+    guard let domain: Domain = try? valueMaps.value(for: "domain"),
+      let primaryType: String = try? valueMaps.value(for: "primaryType"),
+      let messageObject: [String: Any] = try? valueMaps.value(for: "message"),
+      let message = try? MessageObject(from: messageObject, typeName: primaryType, types: types) else {
+      throw EtherKitError.web3Failure(reason: .parsingFailure)
+    }
+
+    return TypedData(types: types, domain: domain, primaryType: primaryType, message: message)
+  }
+
+  public func getDomainSeparator() throws -> Data {
+    let typeHash = try types.getTypeHash(for: "EIP712Domain")
+    let encodedData = try domain.encodeData()
+    return (typeHash + encodedData).sha3(.keccak256)
+  }
+
+  public func getTypeHash() throws -> Data {
+    return try types.getTypeHash(for: primaryType)
+  }
+
+  public func encodeData() throws -> Data {
+    return try message.encodeData(typeName: primaryType, types: types)
+  }
+
+  public func getHashStruct() throws -> Data {
```

Comment:
You're not really getting a HashStruct here, you're getting HashData or just getHash

Reply:
Its called this because thats literally what its called in the EIP and I tried to keep the function names as close to the EIP as possible.